api/triage_metadata.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2019 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package api
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/go-github/v28/github"
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	sourceOwner   = "web-platform-tests"
+	sourceRepo    = "wpt-metadata"
+	commitMessage = "Commit New Metadata"
+	commitBranch  = "auto-triage-branch" + generateRandomInt()
+	baseBranch    = "master"
+	prRepoOwner   = sourceOwner
+	prRepo        = sourceRepo
+	prBranch      = baseBranch
+	prSubject     = "Automatically Triage New Metadata"
+	prDescription = "PR for metadata triaged through /api/metadata/triage endpoint. See <insert a doc> for more information about how to use this service."
+)
+
+type triageMetadata struct {
+	ctx context.Context
+	metadataGithub
+	logger     shared.Logger
+	httpClient *http.Client
+}
+
+type metadataGithub struct {
+	githubClient *github.Client
+	authorName   string
+	authorEmail  string
+}
+
+func (tm triageMetadata) getRef() (ref *github.Reference, err error) {
+	client := tm.githubClient
+	if ref, _, err = client.Git.GetRef(tm.ctx, sourceOwner, sourceRepo, "refs/heads/"+commitBranch); err == nil {
+		return ref, nil
+	}
+
+	// We consider that an error means the branch has not been found and needs to
+	// be created.
+	if commitBranch == baseBranch {
+		return nil, errors.New("The commit branch does not exist but `-base-branch` is the same as `-commit-branch`")
+	}
+
+	var baseRef *github.Reference
+	if baseRef, _, err = client.Git.GetRef(tm.ctx, sourceOwner, sourceRepo, "refs/heads/"+baseBranch); err != nil {
+		return nil, err
+	}
+	newRef := &github.Reference{Ref: github.String("refs/heads/" + commitBranch), Object: &github.GitObject{SHA: baseRef.Object.SHA}}
+	ref, _, err = client.Git.CreateRef(tm.ctx, sourceOwner, sourceRepo, newRef)
+	return ref, err
+}
+
+// getTree generates the tree to commit based on the given files and the commit
+// of the ref you got in getRef.
```

Comment:
Updated. The API doesn't require ref to work but it wouldn't work without ref in this context. IIUC, without ref, the tree created will only include the changes in triageMetadataMap and delete the rest of content in the repo; see https://developer.github.com/v3/git/trees/#create-a-tree.

I would prefer passing `getRef` instead because (1) ref is shared by `pushCommit` and (2)each method represents the flow of creating a Git Objects - create blobs using files, create a branch out of master, create a tree using blobs and the branch ref, create a commit using tree and the branch ref.